### ISSUE TITLE
Run-time Params: A new type of Params not compile-time

### ISSUE THREAD

**Issue by nietras** [dotnet/BenchmarkDotNet#616]
For many of the benchmarks I do I am not interested in `[Params(...)]` being used as compile time constants. And in fact not even interested in there being a completely separate build for each value. With many params and multiple of these this gets really cumbersome and slow.

What recommendation is there for when you simply want parameters used as run-time values? And do not want a build for each value. Only a benchmark run. All values could be run as part of same process. But I do not want new IL generated per run. So InProcess is not what I want exactly it seems.

Additionally, if such `RunTimeParams` or `DynamicParams` where available one could use these with complex types or even test data generators (defined as dynamic params), without the effort of `IParam`. Something similar to `MemberData` and `InlineData` in XUnit.

**Comment by adamsitnik** [comment]
> not even interested in there being a completely separate build for each value. With many params and multiple of these this gets really cumbersome and slow.

I agree that we could build a single executable for all the `Parameters`. It would reduce the time it takes to run the benchmarks. However, I have recently introduced Parallel builds and now the time it takes to build the benchmarks has been improved. For my machine with 4 cores it used to take 40s to build 4 benchmarks for .NET Core, now it's 14s. If I implement the proposed change it would still take 10s and make our code more complex. As for now I dont' think it's worth it. 

> What recommendation is there for when you simply want parameters used as run-time values? And do not want a build for each value. Only a benchmark run. All values could be run as part of same process. But I do not want new IL generated per run. So InProcess is not what I want exactly it seems.

InProcess is the best thing we have for your scenario. It will be the same process, the value will be reused (no need to have compile time constant if you use `IParam`) but the IL will be generated `n` times, where `n` is the number of parameters. However, it should not be a problem in terms of perf.

> Something similar to MemberData and InlineData in XUnit.

We have [ParamsSource](http://benchmarkdotnet.org/Advanced/Params.htm#paramssource) for that.

**Comment by adamsitnik** [comment]
@nietras example:

```cs
class Program
{
    static void Main(string[] args) => BenchmarkRunner.Run<Sample>();
}

[InProcess]
public class Sample
{
    [ParamsSource(nameof(Parameters))]
    public int[] Field;

    [Benchmark]
    public int Sum()
    {
        int sum = 0;
        for (int i = 0; i < Field.Length; i++)
            sum += Field[i];

        return sum;
    }

    public IEnumerable<IParam> Parameters()
    {
        yield return new CustomParam(Enumerable.Range(0, 10).ToArray());
        yield return new CustomParam(Enumerable.Range(0, 100).ToArray());
        yield return new CustomParam(Enumerable.Range(0, 1000).ToArray());
    }

    public class CustomParam : IParam
    {
        public CustomParam(int[] value) => Value = value;

        public object Value { get; }

        public string DisplayText => (Value as int[])?.Length.ToString();

        public string ToSourceCode() => throw new InvalidOperationException("Not used");
    }
}
```

**Comment by nietras** [comment]
> As for now I dont' think it's worth it.

Perfectly understandable, `InProcessToolchain.Instance` does work and helps a lot. And I did use it to define my benchmarks see:

https://github.com/DotNetCross/Sorting/blob/span-sort/tests/DotNetCross.Sorting.Benchmarks/SortBench.cs
https://github.com/DotNetCross/Sorting/blob/span-sort/tests/DotNetCross.Sorting.Benchmarks/SpanFillerParam.cs
https://github.com/DotNetCross/Sorting/blob/span-sort/tests/DotNetCross.Sorting.Benchmarks/SortBenchConfig.cs
https://github.com/DotNetCross/Sorting/blob/span-sort/tests/DotNetCross.Sorting.Benchmarks/Program.cs

> We have ParamsSource for that.

Sorry, probably didn't explain myself too well, I know but this will always be a compile time constant, unless you use `InProcess` and this has it's own set of issues. I think for most cases params are used not to be compile time constants (e.g. trying out a constant buffer size of 32 versus 64) but instead as run-time parameters to methods. When BDN always uses these as compile-time constants I think it skews the benchmarks. In principle the JIT could remove all code for a given constant. It can't do that for a runtime parameter.

Perhaps, you can explain just how the:
```
public string ToSourceCode()
```
is used? I assume only when not doing in process, and that BDN then uses it to generate the C# source code before compiling, right? So when using InProcess the `Value` is used?

**Comment by adamsitnik** [comment]
> . When BDN always uses these as compile-time constants I think it skews the benchmarks. In principle the JIT could remove all code for a given constant. It can't do that for a runtime parameter.

Don't worry, we take care of this. You can see the C# code file that we generete by applying `[KeepBenchmarkFiles]` attribute to your class with benchmarks. We are not using consts or readonly fields, just public fields or public properties initalized in ctor of class which derives from your class with benchmarks.  Speaking of arguments I implemented suport for it, the PR #577 is waiting for @AndreyAkinshin initial verification. Once Andrey approves the PR I am going to implement the missing part (disassembler & inprocess suport)

> I assume only when not doing in process, and that BDN then uses it to generate the C# source code before compiling, right? So when using InProcess the Value is used?

exactly!

**Comment by AndreyAkinshin** [comment]
@adamsitnik ,

>  Speaking of arguments I implemented support for it, the PR #577 is waiting for @AndreyAkinshin initial verification

Sorry, I thought we were stuck on the InProcess support. The current implementation looks good to me. I still I'm still a little worried about nanobenchmarks. However, I guess that there is no a "reliable" approach which always prints "correct" results.
Imagine a method like
```cs
[Arguments(1, 2, 3, 4, 5)]
public void Calc(double a, double b, double c, double d, double e) => a * b * c * d * e;
```
In real life, the `Calc(1, 2, 3, 4, 5)` can be inlined; so we get a compile-time constant. The execution time of such method is `0`, but BenchmarkDotNet will probably print a positive value. In this sense, the `Params` approach is more "inlining-resistant".

However, it can be a problem only for nanobenchmarks (when a method takes a few nanoseconds). You never know what exactly a user want to get (in most cases, users are also don't know). If a benchmark takes more than `1us`, it shouldn't be a problem (I hope).

Here is my suggestion: let's finish the current `Arguments` approach. As soon as it's ready, I'm going to start another crusade on nanobenchmarks: I will check arguments and params with different toolchains and different runtimes (.NET Framework, .NET Core, Mono). In case of any confusing results (when it's really impossible to honestly evaluate performance without knowledge about how a user is going to call it), we can just print a warning like "It's not recommended to use arguments for nanobenchmarks, please, rewrite the benchmarks as follows: bla-bla-bla".

What do you think?

**Comment by adamsitnik** [comment]
@AndreyAkinshin sounds good to me!
